File: task/linkedlist/fastandslowpointers/MiddleOfTheLinkedList.java
package linkedlist.fastandslowpointers;

import linkedlist.ListNode;

public class MiddleOfTheLinkedList {

    public static void main(String[] args) {
        System.out.println(middleNode(new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, new ListNode(6))))))));
//        System.out.println(middleNode(new ListNode(1, new ListNode(2))));
    }

    public static ListNode middleNode(ListNode head) {
        ListNode fast = head, slow = head;

        while (fast != null) {
            if (fast.next == null) {
                return slow.next;
            }
            fast = fast.next.next;
            slow = slow.next;
        }

        return slow;
    }
}

File: python/week2/MiddleOfTheLinkedList.py
from typing import Optional

from python.ListNode import ListNode


class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head.next: return head

        slow, fast = head, head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        return slow


if __name__ == '__main__':
    res2 = Solution().middleNode(ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5, ListNode(6)))))))
    res1 = Solution().middleNode(ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5))))))
    print()
